File: 2/PDOO/NapakalakiJava/src/napakalakiGame/CultistPlayer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalakiGame;

import java.util.Random;

/**
 *
 * @author alejandro
 */
public class CultistPlayer extends Player{
    private static int totalCultisPlayers=0;
    private Cultist myCultistCard;
    
    public static int getTotalCultistPlayers(){
        return totalCultisPlayers;
    }
    
    public CultistPlayer(Player p,Cultist c){
        super(p);
        myCultistCard=c;
        totalCultisPlayers++;
    }
    
    protected int getCombatLevel(){
        int res=(int)(super.getCombatLevel()*1.2)+myCultistCard.getGainedLevels()*totalCultisPlayers;
        return res;
    }
    
    protected Treasure giveMeATreasure(){
         Random rnd=new Random();
        int a=(int)(rnd.nextDouble() * getVisibleTreasures().size());
        return getVisibleTreasures().remove(a);
    }
    
    public boolean canYouGiveMeATreasure(){
        return !getVisibleTreasures().isEmpty();
    }
    
    protected int getOponentLevel(Monster m){
        return m.getCombatLevelAgainstCultisPlayer();
    }


    protected boolean shouldConvert(){
        return false;
    }
    
}

File: 2/PDOO/NapakalakiRuby/lib/cultist_player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'player'
require_relative 'cultist'
module NapakalakiGame
  class CultistPlayer < Player
    @@totalCultistPlayers = 0
    def initialize(p,c)
      @myCultistCard=c
      @@totalCultistPlayers += 1
      self.copy(p)
    end
    
    def self.totalCultistPlayers
      @@totalCultistPLayers
    end
    
    protected
    def getCombatLevel
      (Player.instance_method(:getCombatLevel).bind(self).call*1.2+@myCultistCard.gainedLevels*@@totalCultistPlayers).to_i
    end
    
    def getOponentLevel(m)
      m.levelChangeAgainstCultistPlayer
    end
    
    def shouldConvert
      false
    end
    
    protected
    def giveMeATreasure
      @visibleTreasures.delete_at(rand(@visibleTreasures.size))
    end
    
    def canYouGiveMeATreasure
      !@visibleTreasures.empty?
    end
    
  end
end
